http-webhook.md

```diff
@@ -0,0 +1,332 @@
+formal definition for Web Hooks. This specification aims to provide such a
+definition for use with [CNCF Cloud Events][CE], but is considered generally
+usable beyond the scope of Cloud Events.
+
+## Status of this document
+
+This document is a working draft.
+
+## Table of Contents
+
+1. [Introduction](#1-introduction)
+- 1.1. [Conformance](#11-conformance)
+- 1.2. [Relation to HTTP](#12-relation-to-http)
+2. [Delivering notifications](#2-delivering-notifications)
+3. [Authorization](#3-authorization)
+4. [Abuse Protection](#4-abuse-protection)
+5. [References](#5-references)
+
+## 1. Introduction
+
+"Web hooks" are a popular pattern to deliver notifications between applications
+and via HTTP endpoints. Applications that make notifications available, allow
+for other applications to register an HTTP endpoint to which notifications are
+delivered once available.
+
+This specification defines a HTTP method by how notifications are delivered by
+the sender, an authorization model for event delivery to protect the receiver,
+and a registration handshake that protects the sender from being abused for
+flooding arbitrary HTTP sites with requests.
+
+### 1.1. Conformance
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC2119][RFC2119].
+
+### 1.2. Relation to HTTP
+
+This specification prescribes rules constraining the use and handling of
+specific [HTTP methods][RFC7231-Section-4] and headers.
+
+This specification also applies equivalently to HTTP/2 ([RFC7540][RFC7540]),
+which is compatible with HTTP 1.1 semantics.
+
+## 2. Delivering notifications
+
+### 2.1. Delivery request
+
+Notifications are delivered using a HTTP request. The response indicates the
+resulting status of the delivery.
+
+HTTP-over-TLS (HTTPS) [RFC2818][RFC2818] MUST be used for the connection.
+
+The HTTP method for the delivery request MUST be [POST][POST].
+
+The [`Content-Type`][Content-Type] header MUST be carried and the request MUST
+carry a notification payload of the given content type. Requests without
+payloads, e.g. where the notification is entirely expressed in HTTP headers,
+are not permitted by this specification.
+
+This specification does not further constrain the content of the notification,
+and it also does not prescribe the [HTTP target resource][RFC7230-Section-5-1]
+that is used for delivery.
+
+If the delivery target supports and requires [Abuse Protection][#4-abuse-protection],
+the delivery request MUST include the `Origin` header. The `Origin` header value
+is a DNS name expression that identifies the sending system.
+
+### 2.2. Delivery response
+
+The delivery response MAY contain a payload providing detail status
+information in the case of handling errors. This specification does not attempt
+to define such a payload.
+
+The response MUST NOT use any of the [3xx HTTP Redirect status codes][3xx] and
```

Comment:
Curious what's the rationale for this constraint? To reduce scope, make spec initially more focused?

Reply:
Protection for the sender. I assume that externally supplied targets may be malicious and may want to send the sender on a neverending journey through redirects, so I disallow those outright. Also, the abuse protection only probes this site and not a site that is being redirected to. With redirects, you can basically sidestep the abuse mechanism with a static site that simply redirects all traffic to a DoS target. So this is security related from both angles.

Reply:
This makes sense to me. I imagine there's other status codes that don't make sense, and headers. For example does it makes sense to return an ETag? If we assume the webhook use case is for a response to some event, then ETags would serve no purpose.

Reply:
For that matter caching headers in general wouldn't make sense.

Reply:
While I understand the concerns, isn't this kind of harsh? I come across websites all the time that issue redirects for mindless reasons, like a missing "/" at the end of the URL, and most http clients just blindly handle it for you. Any reason we can't let the sender decide whether they want to support redirects or not?